lib/Sema/TypeCheckType.cpp

```diff
@@ -1065,8 +1065,11 @@ static Type diagnoseUnknownType(TypeResolution resolution,
           methodDecl->getDeclContext() == dc->getParentForLookup();
         bool isPropertyOfClass = insideClass &&
           options.is(TypeResolverContext::PatternBindingDecl);
+        bool isTypeAliasInClass = insideClass &&
+          options.is(TypeResolverContext::TypeAliasDecl);
 
-        if (((!insideClass || !declaringMethod) && !isPropertyOfClass &&
+        if (((!insideClass || !declaringMethod) &&
+             !isPropertyOfClass && !isTypeAliasInClass &&
```

Comment:
I think this set of conditions is pretty much equivalent to `TypeResolutionContext::DynamicSelfResult`, with the added affordance for `Self` inside local context (eg, TypeLocs in method bodies). Take a look at this code snippet in resolveTopLevelIdentTypeComponent():

```
  // Dynamic 'Self' in the result type of a function body.
  if (options.getBaseContext() == TypeResolverContext::DynamicSelfResult &&
      comp->getIdentifier() == ctx.Id_Self) {
```

Can all this logic be moved from `diagnoseUnknownType()` to `resolveTopLevelIdentTypeComponent()`? and combined with the existing code that returns a DynamicSelfType there? You'll need to change it to only build a DynamicSelfType when inside a class, and relax the condition to also allow Self anywhere in structs and enums, and in function bodies of class methods.

Reply:
I'll take a look tomorrow. I've been looking at allowing read only properties and come up against a problem. fn->setDynamicSelf(true) is called on main() rather than the accessor. Any ideas? dc->getInnermostMethodContext() running astray somewhere?

Reply:
I've made another commit of the where things are at this stage. Consolidating to `resolveTopLevelIdentTypeComponent()`seems difficult and getting read only properties and subscripts also difficult due to the main function being marked as being `hasDynamicSelf` rather than the accessor causing an assertion failure generating the getter let alone the setter.

Reply:
Hi @slavapestov, I've pushed another commit that centralises the now very messy SE0068 code in a function and moves it's usage off the diagnostic path. I also have some code which detects if a property or subscript is settable:
```C++
// Causes: Assertion failed: (hasSelfMetadataParam() && "This method can only be called if the " "SILFunction has a self-metadata parameter"), function getSelfMetadataArgument, file /Volumes/Elements/swift-self/swift/include/swift/SIL/SILFunction.h, line 955.
    if (isMutablePropertyOrSubscriptOfClass) {
      if (auto prop = dyn_cast_or_null<ValueDecl>
          (dc->getInnermostDeclarationDeclContext()))
        if (!prop->isSettable(dc))
          isMutablePropertyOrSubscriptOfClass = false;
    }
```
At this point I'm stuck. if you uncomment this code in the commit, you can declare an unsettable property on a class but SILGen breaks somehow. I've got it boxed into:
```Swift
class A1 {
  required init() {}
  func copy() -> Self {
    let copy = Self.init()
    return copy
  }

  var copied: Self {
    let copy = Self.init()
    return copy
  }
  subscript (i: Int) -> Self {
    get {
      return Self.init()
    }
//    set(newValue) {
//    }
  }
}
```
The following is fine:
```Swift
print(A1()[0] as A1)
print(A1().copied as A1)
```
The following gives the assertion failure I mention in the commit:
```Swift
print(A1()[0])
print(A1().copied)
```
The assertion failure is in codegen of the main function which is presented with a "concreteFormalType" which is a dynamic_self_type in SILGenConvert.cpp, when SGF.getFunction().hasSelfMetadataParam() is false. Ring any bells? Full trace is:
```
Assertion failed: (hasSelfMetadataParam() && "This method can only be called if the " "SILFunction has a self-metadata parameter"), function getSelfMetadataArgument, file /Volumes/Elements/swift-self/swift/include/swift/SIL/SILFunction.h, line 955.
Stack dump:
0.	Program arguments: /Volumes/Elements/swift-self/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift -frontend -interpret j.swift -enable-objc-interop -sdk /Applications/Xcode1023.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -O -color-diagnostics -module-name j
0  swift                    0x000000010f0daf25 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift                    0x000000010f0da1e5 llvm::sys::RunSignalHandlers() + 85
2  swift                    0x000000010f0db508 SignalHandler(int) + 264
3  libsystem_platform.dylib 0x00007fff6cadaf5a _sigtramp + 26
4  swift                    0x000000010f0c3572 llvm::raw_ostream::write(char const*, unsigned long) + 466
5  libsystem_c.dylib        0x00007fff6c8781ae abort + 127
6  libsystem_c.dylib        0x00007fff6c8401ac basename_r + 0
7  swift                    0x000000010becc7c4 swift::SILFunction::getSelfMetadataArgument() const + 100
8  swift                    0x000000010bebe7a7 collectTypeDependentOperands(llvm::SmallVectorImpl<swift::SILValue>&, swift::SILOpenedArchetypesState&, swift::SILFunction&, swift::CanType, swift::SubstitutionMap) + 663
9  swift                    0x000000010bec788e swift::InitExistentialAddrInst::create(swift::SILDebugLocation, swift::SILValue, swift::CanType, swift::SILType, llvm::ArrayRef<swift::ProtocolConformanceRef>, swift::SILFunction*, swift::SILOpenedArchetypesState&) + 94
10 swift                    0x000000010b9bdfbb swift::Lowering::SILGenBuilder::createInitExistentialAddr(swift::SILLocation, swift::SILValue, swift::CanType, swift::SILType, llvm::ArrayRef<swift::ProtocolConformanceRef>) + 267
11 swift                    0x000000010b9da21a (anonymous namespace)::ExistentialInitialization::getAddressForInPlaceInitialization(swift::Lowering::SILGenFunction&, swift::SILLocation) + 682
12 swift                    0x000000010b9def91 swift::Lowering::SingleBufferInitialization::copyOrInitValueInto(swift::Lowering::SILGenFunction&, swift::SILLocation, swift::Lowering::ManagedValue, bool) + 49
13 swift                    0x000000010b978249 (anonymous namespace)::ScalarResultPlan::finish(swift::Lowering::SILGenFunction&, swift::SILLocation, swift::CanType, llvm::ArrayRef<swift::Lowering::ManagedValue>&) + 857
14 swift                    0x000000010b990ad7 swift::Lowering::SILGenFunction::emitApply(std::__1::unique_ptr<swift::Lowering::ResultPlan, std::__1::default_delete<swift::Lowering::ResultPlan> >&&, swift::Lowering::ArgumentScope&&, swift::SILLocation, swift::Lowering::ManagedValue, swift::SubstitutionMap, llvm::ArrayRef<swift::Lowering::ManagedValue>, swift::Lowering::CalleeTypeInfo const&, swift::Lowering::ApplyOptions, swift::Lowering::SGFContext) + 1495
15 swift                    0x000000010b995327 (anonymous namespace)::CallEmission::apply(swift::Lowering::SGFContext) + 3431
16 swift                    0x000000010b99daca swift::Lowering::SILGenFunction::emitGetAccessor(swift::SILLocation, swift::SILDeclRef, swift::SubstitutionMap, swift::Lowering::ArgumentSource&&, bool, bool, swift::Lowering::PreparedArguments&&, swift::Lowering::SGFContext, bool) + 682
17 swift                    0x000000010ba2c467 (anonymous namespace)::GetterSetterComponent::get(swift::Lowering::SILGenFunction&, swift::SILLocation, swift::Lowering::ManagedValue, swift::Lowering::SGFContext) && + 311
18 swift                    0x000000010ba1f1f4 swift::Lowering::SILGenFunction::emitLoadOfLValue(swift::SILLocation, swift::Lowering::LValue&&, swift::Lowering::SGFContext, bool) + 788
19 swift                    0x000000010b9ffe69 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 18889
20 swift                    0x000000010b9f0559 swift::Lowering::SILGenFunction::emitRValueAsSingleValue(swift::Expr*, swift::Lowering::SGFContext) + 57
21 swift                    0x000000010ba0a7d5 swift::Lowering::ManagedValue llvm::function_ref<swift::Lowering::ManagedValue (swift::Lowering::SGFContext)>::callback_fn<(anonymous namespace)::RValueEmitter::visitErasureExpr(swift::ErasureExpr*, swift::Lowering::SGFContext)::$_8>(long, swift::Lowering::SGFContext) + 101
22 swift                    0x000000010b9da64e void llvm::function_ref<void (swift::SILValue)>::callback_fn<swift::Lowering::SILGenFunction::emitExistentialErasure(swift::SILLocation, swift::CanType, swift::Lowering::TypeLowering const&, swift::Lowering::TypeLowering const&, llvm::ArrayRef<swift::ProtocolConformanceRef>, swift::Lowering::SGFContext, llvm::function_ref<swift::Lowering::ManagedValue (swift::Lowering::SGFContext)>, bool)::$_6>(long, swift::SILValue) + 158
23 swift                    0x000000010b9bfc0a swift::Lowering::SILGenBuilder::bufferForExpr(swift::SILLocation, swift::SILType, swift::Lowering::TypeLowering const&, swift::Lowering::SGFContext, llvm::function_ref<void (swift::SILValue)>) + 154
24 swift                    0x000000010b9d4e48 swift::Lowering::SILGenFunction::emitExistentialErasure(swift::SILLocation, swift::CanType, swift::Lowering::TypeLowering const&, swift::Lowering::TypeLowering const&, llvm::ArrayRef<swift::ProtocolConformanceRef>, swift::Lowering::SGFContext, llvm::function_ref<swift::Lowering::ManagedValue (swift::Lowering::SGFContext)>, bool) + 1672
25 swift                    0x000000010ba0041a swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 20346
26 swift                    0x000000010b9fb64a swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 426
27 swift                    0x000000010b9ed2ff swift::Lowering::SILGenFunction::emitExprInto(swift::Expr*, swift::Lowering::Initialization*, llvm::Optional<swift::SILLocation>) + 303
28 swift                    0x000000010b9717c9 swift::Lowering::ArgumentSource::forwardInto(swift::Lowering::SILGenFunction&, swift::Lowering::Initialization*) && + 217
29 swift                    0x000000010b971e74 swift::Lowering::ArgumentSource::forwardInto(swift::Lowering::SILGenFunction&, swift::Lowering::AbstractionPattern, swift::Lowering::Initialization*, swift::Lowering::TypeLowering const&) && + 548
30 swift                    0x000000010b9a304b (anonymous namespace)::ArgEmitter::emit(swift::Lowering::ArgumentSource&&, swift::Lowering::AbstractionPattern) + 3723
31 swift                    0x000000010b9afa48 (anonymous namespace)::ArgEmitter::emitTopLevel(swift::Lowering::ArgumentSource&&, swift::Lowering::AbstractionPattern) + 5544
32 swift                    0x000000010b9ae1fe (anonymous namespace)::CallSite::emit(swift::Lowering::SILGenFunction&, swift::Lowering::AbstractionPattern, swift::CanTypeWrapper<swift::SILFunctionType>, (anonymous namespace)::ParamLowering&, llvm::SmallVectorImpl<swift::Lowering::ManagedValue>&, llvm::SmallVectorImpl<(anonymous namespace)::DelayedArgument>&, llvm::Optional<swift::ForeignErrorConvention> const&, swift::ImportAsMemberStatus) && + 1006
33 swift                    0x000000010b9ad90c (anonymous namespace)::CallEmission::emitArgumentsForNormalApply(swift::CanTypeWrapper<swift::FunctionType>&, swift::Lowering::AbstractionPattern&, swift::CanTypeWrapper<swift::SILFunctionType>, llvm::Optional<swift::ForeignErrorConvention> const&, swift::ImportAsMemberStatus, llvm::SmallVectorImpl<swift::Lowering::ManagedValue>&, llvm::Optional<swift::SILLocation>&, swift::CanTypeWrapper<swift::FunctionType>&) + 1772
34 swift                    0x000000010b9951a5 (anonymous namespace)::CallEmission::apply(swift::Lowering::SGFContext) + 3045
35 swift                    0x000000010b9944bf swift::Lowering::SILGenFunction::emitApplyExpr(swift::Expr*, swift::Lowering::SGFContext) + 2111
36 swift                    0x000000010b9fb4f0 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 80
37 swift                    0x000000010b9f0a81 swift::Lowering::SILGenFunction::emitIgnoredExpr(swift::Expr*) + 1137
38 swift                    0x000000010b9890e4 swift::Lowering::SILGenModule::visitTopLevelCodeDecl(swift::TopLevelCodeDecl*) + 564
39 swift                    0x000000010b989906 swift::Lowering::SILGenModule::emitSourceFile(swift::SourceFile*) + 822
40 swift                    0x000000010b98a908 swift::SILModule::constructSIL(swift::ModuleDecl*, swift::SILOptions&, swift::FileUnit*) + 376
41 swift                    0x000000010b98adf0 swift::performSILGeneration(swift::ModuleDecl*, swift::SILOptions&) + 16
42 swift                    0x000000010b691f02 performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 10258
43 swift                    0x000000010b68e692 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2978
44 swift                    0x000000010b635578 main + 696
45 libdyld.dylib            0x00007fff6c7cc015 start + 1
46 libdyld.dylib            0x000000000000000b start + 2474852343
Abort trap: 6

Reply:
Let's not worry about allowing Self on read-only properties and subscripts in this PR. I just wanted you to add a test that this was properly diagnosed. I can help you get it working if you're interested but it's a bit more involved and requires changing some code in the constraint solver (search for getTypeOfMemberReference() and buildMemberRef()).